File: sdk-py/src/capsule/builder.py
"""CapsuleBuilder — plain and encrypted capsule build paths. Mirrors sdk/src/builder.js."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from datetime import UTC

from .canonical import bytes_to_hex, hex_to_bytes, jcs, sha256_hex
from .chain import build_chain_events, events_to_jsonl, first_and_entry_hash
from .crypto import (
    chacha20_poly1305_encrypt,
    generate_x25519,
    hkdf_sha256,
    random_key32,
    random_nonce12,
    x25519_dh,
)
from .envelope import build_envelope, sign_envelope
from .manifest import (
    build_content_index,
    build_manifest,
    compute_capsule_id,
    manifest_bytes,
    manifest_hash,
)
from .pith import compress_event_payload
from .zip_io import pack_zip

_SKILL_ID_RE = re.compile(r"^[a-zA-Z0-9_-]+$")


@dataclass
class _SkillEntry:
    json: dict | None
    markdown: str | None
    signed: bool


class CapsuleBuilder:
    def __init__(
        self,
        *,
        originator: dict,
        participants: list[dict] | None = None,
        created_at: str | None = None,
        pith: bool = True,
    ) -> None:
        if not isinstance(originator, dict) or not isinstance(originator.get("public_key"), str):
            raise ValueError("originator.public_key (hex) required")
        self.originator = {
            "public_key": originator["public_key"],
            "label": originator.get("label", ""),
        }
        self.participants = participants or []
        self.created_at = created_at or _now_no_fractional()
        self.program_md: str | None = None
        self.agents_md: str | None = None
        self.skills: dict[str, _SkillEntry] = {}
        self.payload: dict[str, bytes] = {}
        self.bare_events: list[dict] = []
        self.pith = bool(pith)

    def set_program(self, md: str) -> CapsuleBuilder:
        self.program_md = md
        return self

    def set_agents(self, md: str) -> CapsuleBuilder:
        self.agents_md = md
        return self

    def add_skill(
        self,
        id: str,
        *,
        json: dict | None = None,
        markdown: str | None = None,
        signed: bool = False,
    ) -> CapsuleBuilder:
        if not isinstance(id, str) or not _SKILL_ID_RE.match(id):
            raise ValueError(f"invalid skill id: {id}")
        if id == "decryption":
            raise ValueError("'decryption' is reserved for encryption metadata; not a skill")
        self.skills[id] = _SkillEntry(json=json, markdown=markdown, signed=bool(signed))
        return self

    def add_payload(self, path: str, data: bytes) -> CapsuleBuilder:
        if not isinstance(path, str) or not path.startswith("payload/"):
            raise ValueError(f"payload path must start with 'payload/': {path}")
        self.payload[path] = bytes(data)
        return self

    def append_event(self, event: dict, *, pith: bool | None = None) -> CapsuleBuilder:
        for required in ("actor", "kind", "action", "target"):
            if required not in event:
                raise ValueError(f"event requires {required}")
        apply_pith = self.pith if pith is None else (self.pith and pith)
        raw_payload = event.get("payload", {})
        payload = compress_event_payload(raw_payload) if apply_pith else raw_payload
        bare = {
            "actor": event["actor"],
            "kind": event["kind"],
            "action": event["action"],
            "target": event["target"],
            "timestamp": event.get("timestamp", self.created_at),
            "payload": payload,
        }
        if "untrusted_payload_fields" in event:
            bare["untrusted_payload_fields"] = event["untrusted_payload_fields"]
        self.bare_events.append(bare)
        return self

    def seal(
        self,
        *,
        signers: list[dict],
        signed_at: str,
        recipients: list[bytes] | None = None,
    ) -> bytes:
        if not signers:
            raise ValueError("seal requires at least one signer")
        if not signed_at:
            raise ValueError("seal requires signed_at")

        if self.program_md is None:
            self.program_md = "# Program\n"
        if not self.bare_events:
            self.bare_events.append(
                {
                    "actor": "system:host",
                    "kind": "observation",
                    "action": "session_ended",
                    "target": "capsule",
                    "timestamp": signed_at,
                    "payload": {"note": "host emitted backstop event before seal"},
                }
            )

        # Chain
        events = build_chain_events(self.bare_events)
        first_event_hash, entry_hash = first_and_entry_hash(events)
        events_jsonl = events_to_jsonl(events)

        # Inner files
        inner: dict[str, bytes] = {
            "program.md": self.program_md.encode("utf-8"),
            "chain/events.jsonl": events_jsonl,
        }
        if self.agents_md is not None:
            inner["agents.md"] = self.agents_md.encode("utf-8")
        skill_trust: dict[str, str] = {}
        for sid, entry in self.skills.items():
            if entry.json is not None:
                inner[f"skills/{sid}/skill.json"] = json.dumps(
                    entry.json, indent=2, ensure_ascii=False
                ).encode("utf-8")
            if entry.markdown is not None:
                inner[f"skills/{sid}/SKILL.md"] = entry.markdown.encode("utf-8")
            skill_trust[sid] = "signed" if entry.signed else "unsigned"
        inner.update(self.payload)

        # Manifest
        originator_pub_raw = hex_to_bytes(self.originator["public_key"])
        capsule_id = compute_capsule_id(originator_pub_raw, first_event_hash)

        # ---- Plain path ----
        if not recipients:
            content_index = build_content_index(inner)
            manifest = build_manifest(
                originator=self.originator,
                participants=self.participants,
                content_index=content_index,
                first_event_hash=first_event_hash,
                skill_trust=skill_trust,
                encryption=None,
                created_at=self.created_at,
            )
            manifest["id"] = capsule_id
            mf_hash = manifest_hash(manifest)

            envelope = build_envelope(
                capsule_id=capsule_id,
                first_event_hash=first_event_hash,
                entry_hash=entry_hash,
                manifest_hash=mf_hash,
                content_index_hash=content_index["index_hash"],
                encrypted_blob_hash=None,
                cipher="none",
                signed_at=signed_at,
            )
            sign_envelope(envelope, signers)

            all_files = dict(inner)
            all_files["manifest.json"] = manifest_bytes(manifest)
            all_files["provenance/envelope.json"] = json.dumps(
                envelope, indent=2, ensure_ascii=False
            ).encode("utf-8")
            return pack_zip(all_files)

        # ---- Encrypted path ----

        # 3a) Build inner ZIP
        inner_content_index = build_content_index(inner)
        inner_manifest = build_manifest(
            originator=self.originator,
            participants=self.participants,
            content_index=inner_content_index,
            first_event_hash=first_event_hash,
            skill_trust=skill_trust,
            encryption=None,
            created_at=self.created_at,
        )
        inner_manifest["id"] = capsule_id
        inner_mf_hash = manifest_hash(inner_manifest)

        inner_envelope = build_envelope(
            capsule_id=capsule_id,
            first_event_hash=first_event_hash,
            entry_hash=entry_hash,
            manifest_hash=inner_mf_hash,
            content_index_hash=inner_content_index["index_hash"],
            encrypted_blob_hash=None,
            cipher="none",
            signed_at=signed_at,
        )
        sign_envelope(inner_envelope, signers)

        inner_all_files = dict(inner)
        inner_all_files["manifest.json"] = manifest_bytes(inner_manifest)
        inner_all_files["provenance/envelope.json"] = json.dumps(
            inner_envelope, indent=2, ensure_ascii=False
        ).encode("utf-8")
        inner_zip_bytes = pack_zip(inner_all_files)

        # 3b) Encrypt inner ZIP
        content_key = random_key32()
        content_nonce = random_nonce12()

        aad_obj = {
            "capsule_id": capsule_id,
            "cipher": "ChaCha20-Poly1305",
            "first_event_hash": first_event_hash,
            "originator_public_key": self.originator["public_key"],
            "version": "0.6",
        }
        aad = jcs(aad_obj)
        content_enc = chacha20_poly1305_encrypt(content_key, content_nonce, aad, inner_zip_bytes)
        encrypted_blob_hash = sha256_hex(content_enc)

        # 3c) Build recipient bundles
        _content_nonce_hex, decryption_meta = _build_decryption_metadata(
            content_key, content_nonce, recipients
        )

        # 3d) Outer files
        outer_sidecars: dict[str, bytes] = {
            "skills/decryption/decryption.json": json.dumps(
                decryption_meta, indent=2, ensure_ascii=False
            ).encode("utf-8"),
            "content.enc": content_enc,
        }

        outer_content_index = build_content_index(outer_sidecars)

        outer_manifest = build_manifest(
            originator=self.originator,
            participants=self.participants,
            content_index=outer_content_index,
            first_event_hash=first_event_hash,
            skill_trust={},
            encryption={
                "metadata_path": "skills/decryption/decryption.json",
                "cipher": "ChaCha20-Poly1305",
            },
            created_at=self.created_at,
        )
        outer_manifest["id"] = capsule_id
        outer_mf_hash = manifest_hash(outer_manifest)

        outer_envelope = build_envelope(
            capsule_id=capsule_id,
            first_event_hash=first_event_hash,
            entry_hash=entry_hash,
            manifest_hash=outer_mf_hash,
            content_index_hash=outer_content_index["index_hash"],
            encrypted_blob_hash=encrypted_blob_hash,
            cipher="ChaCha20-Poly1305",
            signed_at=signed_at,
        )
        sign_envelope(outer_envelope, signers)

        outer_all_files = dict(outer_sidecars)
        outer_all_files["manifest.json"] = manifest_bytes(outer_manifest)
        outer_all_files["provenance/envelope.json"] = json.dumps(
            outer_envelope, indent=2, ensure_ascii=False
        ).encode("utf-8")
        return pack_zip(outer_all_files)


def _build_decryption_metadata(
    content_key: bytes,
    content_nonce: bytes,
    recipients: list[bytes],
) -> tuple[str, dict]:
    """Build the decryption.json dict for the given recipients.

    Returns ``(content_nonce_hex, decryption_meta_dict)``.
    """
    key_bundles = []
    for recipient_pub in recipients:
        if len(recipient_pub) != 32:
            raise ValueError("recipient public key must be 32 bytes (X25519 raw)")
        eph = generate_x25519()
        shared = x25519_dh(eph.private_key, recipient_pub)
        wrap_key = hkdf_sha256(
            shared,
            recipient_pub,  # salt = recipient public key
            b"capsule-key-wrap-v0.6",
            32,
        )
        wrap_nonce = random_nonce12()
        wrapped_key = chacha20_poly1305_encrypt(wrap_key, wrap_nonce, b"", content_key)
        key_bundles.append(
            {
                "recipient_public_key": bytes_to_hex(recipient_pub),
                "ephemeral_public_key": eph.public_key_hex,
                "wrap_nonce": bytes_to_hex(wrap_nonce),
                "wrapped_key": bytes_to_hex(wrapped_key),
            }
        )

    decryption_meta = {
        "cipher": "ChaCha20-Poly1305",
        "content_nonce": bytes_to_hex(content_nonce),
        "key_bundles": key_bundles,
    }
    return bytes_to_hex(content_nonce), decryption_meta


def _now_no_fractional() -> str:
    from datetime import datetime

    return datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")

File: sdk-js/src/builder.js
// CapsuleBuilder: assembles a v0.6 capsule and seals it.

import { jcs, sha256, sha256Hex } from "./canonical.js";
import {
  buildChainEvents,
  eventsToJsonl,
  firstAndEntryHash,
} from "./chain.js";
import {
  bytesToHex,
  chacha20Poly1305Encrypt,
  generateX25519,
  hkdfSha256,
  hexToBytes,
  randomKey32,
  randomNonce12,
  x25519DH,
} from "./crypto.js";
import {
  buildEnvelope,
  signEnvelope,
} from "./envelope.js";
import {
  buildContentIndex,
  buildManifest,
  computeCapsuleId,
  CONTENT_INDEX_EXCLUDED,
  manifestBytes,
  manifestHash,
} from "./manifest.js";
import { compressEventPayload } from "./pith.js";
import { packZip } from "./zip.js";

export class CapsuleBuilder {
  constructor({ originator, participants = [], createdAt, pith = true }) {
    if (!originator || typeof originator.publicKey !== "string") {
      throw new Error("originator.publicKey (hex) required");
    }
    this.originator = {
      public_key: originator.publicKey,
      label: originator.label ?? "",
    };
    this.participants = participants;
    this.createdAt = createdAt ?? new Date().toISOString().replace(/\.\d+Z$/, "Z");
    this.programMd = null;
    this.agentsMd = null;
    this.skills = new Map(); // id -> { json, markdown, signed }
    this.payload = new Map(); // path -> bytes
    this.bareEvents = [];
    this.pith = pith !== false; // default on; pass {pith:false} to disable
  }

  setProgram(md) {
    this.programMd = md;
    return this;
  }

  setAgents(md) {
    this.agentsMd = md;
    return this;
  }

  addSkill(id, { json, markdown, signed = false }) {
    if (typeof id !== "string" || !/^[a-zA-Z0-9_-]+$/.test(id)) {
      throw new Error(`invalid skill id: ${id}`);
    }
    if (id === "decryption") {
      throw new Error("'decryption' is reserved for encryption metadata; not a skill");
    }
    this.skills.set(id, { json: json ?? null, markdown: markdown ?? null, signed: !!signed });
    return this;
  }

  addPayload(path, bytes) {
    if (!path.startsWith("payload/")) {
      throw new Error(`payload path must start with 'payload/': ${path}`);
    }
    this.payload.set(path, Buffer.from(bytes));
    return this;
  }

  /**
   * Append a chain event. Per-call opt-out: { pith: false } skips
   * payload normalization for this event.
   */
  appendEvent(event, options = {}) {
    if (!event.actor || !event.kind || !event.action || !event.target) {
      throw new Error("event requires actor, kind, action, target");
    }
    const applyPith = options.pith !== false && this.pith;
    const rawPayload = event.payload ?? {};
    const payload = applyPith ? compressEventPayload(rawPayload) : rawPayload;
    this.bareEvents.push({
      actor: event.actor,
      kind: event.kind,
      action: event.action,
      target: event.target,
      timestamp: event.timestamp ?? this.createdAt,
      payload,
      ...(event.untrusted_payload_fields ? { untrusted_payload_fields: event.untrusted_payload_fields } : {}),
    });
    return this;
  }

  /**
   * Seal and emit the capsule bytes.
   *
   * options:
   *   signers:    [{ role, publicKey: 32 bytes, privateKey: 32 bytes }]
   *   recipients: optional [{ publicKey: 32 bytes }] — enables encryption
   *   signedAt:   ISO 8601 UTC string
   */
  async seal({ signers, recipients = [], signedAt }) {
    if (!signers || signers.length === 0) throw new Error("seal requires at least one signer");
    if (!signedAt) throw new Error("seal requires signedAt");
    if (this.programMd == null) this.programMd = "# Program\n";
    if (this.bareEvents.length === 0) {
      // host-emitted backstop event so we never seal an empty chain
      this.bareEvents.push({
        actor: "system:host",
        kind: "observation",
        action: "session_ended",
        target: "capsule",
        timestamp: signedAt,
        payload: { note: "host emitted backstop event before seal" },
      });
    }

    // 1) Build chain
    const events = buildChainEvents(this.bareEvents);
    const { firstEventHash, entryHash } = firstAndEntryHash(events);
    const eventsJsonl = eventsToJsonl(events);

    // 2) Inner files
    const innerFiles = new Map();
    innerFiles.set("program.md", Buffer.from(this.programMd, "utf8"));
    if (this.agentsMd != null) {
      innerFiles.set("agents.md", Buffer.from(this.agentsMd, "utf8"));
    }
    innerFiles.set("chain/events.jsonl", eventsJsonl);
    const skillTrust = {};
    for (const [id, s] of this.skills.entries()) {
      if (s.json != null) {
        innerFiles.set(`skills/${id}/skill.json`, Buffer.from(JSON.stringify(s.json, null, 2), "utf8"));
      }
      if (s.markdown != null) {
        innerFiles.set(`skills/${id}/SKILL.md`, Buffer.from(s.markdown, "utf8"));
      }
      skillTrust[id] = s.signed ? "signed" : "unsigned";
    }
    for (const [path, bytes] of this.payload.entries()) {
      innerFiles.set(path, bytes);
    }

    // 3) Build manifest (without id)
    const originatorPubRaw = hexToBytes(this.originator.public_key);
    const capsuleId = computeCapsuleId(originatorPubRaw, firstEventHash);

    if (recipients.length === 0) {
      // ---- Plain capsule ----
      const contentIndex = buildContentIndex(innerFiles);
      const manifest = buildManifest({
        originator: this.originator,
        participants: this.participants,
        contentIndex,
        firstEventHash,
        skillTrust,
        encryption: null,
        createdAt: this.createdAt,
      });
      manifest.id = capsuleId;
      const mfHash = manifestHash(manifest);

      const envelope = buildEnvelope({
        capsuleId,
        firstEventHash,
        entryHash,
        manifestHash: mfHash,
        contentIndexHash: contentIndex.index_hash,
        encryptedBlobHash: null,
        cipher: "none",
        signedAt,
      });
      signEnvelope(envelope, signers);

      const allFiles = new Map(innerFiles);
      allFiles.set("manifest.json", manifestBytes(manifest));
      allFiles.set("provenance/envelope.json", Buffer.from(JSON.stringify(envelope, null, 2), "utf8"));
      const zipBytes = await packZip(allFiles);
      return zipBytes;
    }

    // ---- Encrypted capsule ----

    // 3a) Pack inner ZIP (still includes manifest with id + content_index covering inner files)
    const innerContentIndex = buildContentIndex(innerFiles);
    const innerManifest = buildManifest({
      originator: this.originator,
      participants: this.participants,
      contentIndex: innerContentIndex,
      firstEventHash,
      skillTrust,
      encryption: null,
      createdAt: this.createdAt,
    });
    innerManifest.id = capsuleId;
    const innerMfHash = manifestHash(innerManifest);

    // The inner envelope is plain — used by L3 to verify the inner package.
    const innerEnvelope = buildEnvelope({
      capsuleId,
      firstEventHash,
      entryHash,
      manifestHash: innerMfHash,
      contentIndexHash: innerContentIndex.index_hash,
      encryptedBlobHash: null,
      cipher: "none",
      signedAt,
    });
    signEnvelope(innerEnvelope, signers);

    const innerAllFiles = new Map(innerFiles);
    innerAllFiles.set("manifest.json", manifestBytes(innerManifest));
    innerAllFiles.set(
      "provenance/envelope.json",
      Buffer.from(JSON.stringify(innerEnvelope, null, 2), "utf8"),
    );
    const innerZipBytes = await packZip(innerAllFiles);

    // 3b) Encrypt inner zip.
    // AAD per spec/envelope.md "Encryption" — version, capsule_id,
    // first_event_hash, originator_public_key, cipher. manifest_hash
    // is intentionally excluded (the outer manifest depends on the
    // encrypted_blob_hash, which depends on this step; the inner
    // content commitment is established at L3 by the inner envelope).
    // Field order is irrelevant — JCS sorts keys lexicographically.
    const contentKey = randomKey32();
    const contentNonce = randomNonce12();

    const aadObj = {
      version: "0.6",
      capsule_id: capsuleId,
      first_event_hash: firstEventHash,
      originator_public_key: this.originator.public_key,
      cipher: "ChaCha20-Poly1305",
    };
    const aad = jcs(aadObj);
    const contentEnc = chacha20Poly1305Encrypt(contentKey, contentNonce, aad, innerZipBytes);
    const encryptedBlobHash = sha256Hex(contentEnc);

    // 3c) Build recipient bundles
    const keyBundles = recipients.map((r) => {
      if (!r.publicKey || r.publicKey.length !== 32) {
        throw new Error("recipient.publicKey must be 32 bytes");
      }
      const eph = generateX25519();
      const shared = x25519DH(eph.privateKey, r.publicKey);
      const wrapKey = hkdfSha256(
        shared,
        r.publicKey,
        Buffer.from("capsule-key-wrap-v0.6", "utf8"),
        32,
      );
      const wrapNonce = randomNonce12();
      const wrappedKey = chacha20Poly1305Encrypt(wrapKey, wrapNonce, Buffer.alloc(0), contentKey);
      return {
        recipient_public_key: bytesToHex(r.publicKey),
        ephemeral_public_key: eph.publicKeyHex,
        wrap_nonce: bytesToHex(wrapNonce),
        wrapped_key: bytesToHex(wrappedKey),
      };
    });

    const decryptionMeta = {
      cipher: "ChaCha20-Poly1305",
      content_nonce: bytesToHex(contentNonce),
      key_bundles: keyBundles,
    };

    // 3d) Outer manifest covers content.enc and decryption metadata
    const outerSidecars = new Map();
    outerSidecars.set(
      "skills/decryption/decryption.json",
      Buffer.from(JSON.stringify(decryptionMeta, null, 2), "utf8"),
    );
    outerSidecars.set("content.enc", contentEnc);
    // Encrypted profile: content.enc is bound by envelope.encrypted_blob_hash,
    // so it is excluded from the content index here.
    const outerContentIndex = buildContentIndex(outerSidecars, CONTENT_INDEX_EXCLUDED);

    const outerManifest = buildManifest({
      originator: this.originator,
      participants: this.participants,
      contentIndex: outerContentIndex,
      firstEventHash,
      skillTrust: {}, // decryption metadata is not a skill
      encryption: {
        metadata_path: "skills/decryption/decryption.json",
        cipher: "ChaCha20-Poly1305",
      },
      createdAt: this.createdAt,
    });
    outerManifest.id = capsuleId;
    const outerMfHash = manifestHash(outerManifest);

    const outerEnvelope = buildEnvelope({
      capsuleId,
      firstEventHash,
      entryHash,
      manifestHash: outerMfHash,
      contentIndexHash: outerContentIndex.index_hash,
      encryptedBlobHash,
      cipher: "ChaCha20-Poly1305",
      signedAt,
    });
    signEnvelope(outerEnvelope, signers);

    const outerAllFiles = new Map(outerSidecars);
    outerAllFiles.set("manifest.json", manifestBytes(outerManifest));
    outerAllFiles.set(
      "provenance/envelope.json",
      Buffer.from(JSON.stringify(outerEnvelope, null, 2), "utf8"),
    );
    return await packZip(outerAllFiles);
  }
}
